lib/client/session.go

```diff
@@ -568,26 +578,28 @@ func (ns *NodeSession) pipeInOut(shell io.ReadWriteCloser) {
 			})
 		}
 		for {
-			n, err := stdin.Read(buf)
-			if err != nil {
-				fmt.Fprintf(ns.stderr, "\r\n%v\r\n", trace.Wrap(err))
+			select {
+			case <-ns.closer.C:
 				return
-			}
-
-			if n > 0 {
-				_, err = shell.Write(buf[:n])
+			default:
+				n, err := stdin.Read(buf)
 				if err != nil {
-					ns.ExitMsg = err.Error()
+					fmt.Fprintf(ns.stderr, "\r\n%v\r\n", trace.Wrap(err))
 					return
 				}
+
+				if n > 0 {
+					_, err = shell.Write(buf[:n])
+					if err != nil {
+						ns.ExitMsg = err.Error()
+						return
+					}
+				}
```

Comment:
in case `stdin.Read` returns `n>0` and `err != nil`, write that last piece of data to `shell`
IOTW, move error handling to the bottom